### DISCUSSION TITLE
How to sample instances of ShapeGroup?

### DISCUSSION THREAD

**Discussion by woAIxuexiSR** [mitsuba-renderer/mitsuba3#990] [Q&A] [answered]
Hi there,

I'm currently working on sampling shapes within my scene using the following code:

```
for shape in scene.shapes():
    ps = shape.sample_position(0, sample2, True)
```

However, when the iteration reaches an instance of a ShapeGroup, specifically:
```
Instance[
  shapegroup = ShapeGroup[
    name = "door_group",
    prim_count = 2956
  ]
  to_world = [[1, 0, 0, 0],
              [0, 1, 0, 0],
              [0, 0, 1, 0],
              [0, 0, 0, 1]],
]
```

I encounter the following error: `RuntimeError: ​[Shape] Instance::sample_position(): not implemented!`
I'm seeking guidance on how to properly sample shapes within a scene when encountering instances of ShapeGroups. Is there a recommended approach or workaround for this issue?

**Comment by njroussel** [comment, marked as answer]
Hi @woAIxuexiSR 

A trivial, albeit inefficient, workaround would be to remove the instancing and just duplicate the shape in the scene description.

I don't there is any other option, this has not been implemented.
If you are interested in working on this, I don't believe that this would be a very difficult task. It should just be a matter of sampling a shape in the group, then sampling a position on that shape and finally applying the instance's transform to that position.

**Reply by woAIxuexiSR** [reply]
Thank for your help.

I want to dynamically apply a transformation to the shape. If I don't use an instance, I can't seem to find the 'to_world' attribute to modify when traversing through the scene parameters.

For example, I have an obj mesh in my scene.
```
<shape type="obj" id="Door">
    <string name="filename" value="models/Mesh006.obj" />
    <ref id="DoorBSDF" />
</shape>
```
I get this scene paramters:
```
Door.faces                                                     UInt              OBJMesh
Door.vertex_positions                                 ∂, D     Float             OBJMesh
Door.vertex_normals                                   ∂, D     Float             OBJMesh
Door.vertex_texcoords                                 ∂        Float             OBJMesh
```
Is there any alternative method other than applying the transformation matrix to each vertex and normal individually?

**Reply by njroussel** [reply]
> Is there any alternative method other than applying the transformation matrix to each vertex and normal individually?

No. We have an example of how to do this, it's only a few lines: https://mitsuba.readthedocs.io/en/latest/src/inverse_rendering/object_pose_estimation.html#Optimizer-and-latent-variables.

Changing the vertex positions of a `Mesh` will make it [recompute its normals](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/render/mesh.cpp#L119) by itself (it's an implementation of [this method](https://dl.acm.org/doi/10.1080/10867651.1998.10487487)). These might not be exactly the normals you originally used. If this is an issue, you can apply a second **separate** update where you only update the `vertex_normals`.
